crates/linker/src/pe.rs

```diff
@@ -0,0 +1,357 @@
+
+            let offset_in_file = self.section_offset_in_file + thunk_start_offset + thunk_offset;
+            let virtual_address = descriptor.original_first_thunk.get(LE) + thunk_offset;
+
+            self.name_by_virtual_address
+                .insert(virtual_address, name.clone());
+            self.address_and_offset
+                .insert(name, (virtual_address, offset_in_file));
+
+            thunk_offset += std::mem::size_of::<ImageThunkData64>() as u32;
+        }
+
+        Ok(())
+    }
+
+    fn new_help(data: &[u8]) -> object::read::Result<Self> {
+        use object::pe;
+        use object::read::pe::ImageNtHeaders;
+        use object::LittleEndian as LE;
+
+        let dos_header = pe::ImageDosHeader::parse(data)?;
+        let mut offset = dos_header.nt_headers_offset().into();
+        let (nt_headers, data_directories) = ImageNtHeaders64::parse(data, &mut offset)?;
+        let sections = nt_headers.sections(data, offset)?;
+
+        let data_dir = match data_directories.get(pe::IMAGE_DIRECTORY_ENTRY_IMPORT) {
+            Some(data_dir) => data_dir,
+            None => internal_error!("No dynamic imports, probably a bug"),
+        };
+
+        let import_va = data_dir.virtual_address.get(LE);
+
+        let (section_data, section_va, offset_in_file) = sections
+            .iter()
+            .find_map(|section| {
+                section
+                    .pe_data_containing(data, import_va)
+                    .map(|(section_data, section_va)| {
+                        (
+                            section_data,
+                            section_va,
+                            section.pointer_to_raw_data.get(LE),
+                        )
+                    })
+            })
+            .expect("Invalid import data dir virtual address");
+
+        let import_table = ImportTable::new(section_data, section_va, import_va);
+
+        let imports_offset_in_section = import_va.wrapping_sub(section_va);
+        let imports_offset_in_file = offset_in_file + imports_offset_in_section;
+
+        let (descriptor, dummy_import_index) = Self::find_roc_dummy_dll(&import_table)?.unwrap();
+
+        let mut this = Self {
+            name_by_virtual_address: Default::default(),
+            address_and_offset: Default::default(),
+            section_virtual_address: section_va,
+            section_offset_in_file: offset_in_file,
+            imports_offset_in_file,
+            dummy_import_index,
+        };
+
+        this.append_roc_imports(&import_table, &descriptor)?;
+
+        Ok(this)
+    }
+}
+
+#[allow(dead_code)]
+fn remove_dummy_dll_import_table(
+    data: &mut [u8],
+    imports_offset_in_file: u32,
+    dummy_import_index: u32,
+) {
+    const W: usize = std::mem::size_of::<ImageImportDescriptor>();
+
+    let start = imports_offset_in_file as usize + W * dummy_import_index as usize;
+    for b in data[start..][..W].iter_mut() {
```

Comment:
I think there is a function that exists to fill a slice that you should be able to use here instead.

Reply:
in the linker code, or in `std`?